include/cantera/kinetics/ReactionRate.h

```diff
@@ -0,0 +1,255 @@
+// at https://cantera.org/license.txt for license and copyright information.
+
+#ifndef CT_REACTIONRATE_H
+#define CT_REACTIONRATE_H
+
+#include "cantera/kinetics/RxnRates.h"
+#include "cantera/base/ctexceptions.h"
+
+namespace Cantera
+{
+
+class Func1;
+class ThermoPhase;
+class AnyMap;
+
+
+//! Abstract base class for reaction rate definitions
+/**
+ * Because this class has no template parameters, derived objects can be
+ * accessed via `shared_ptr<ReactionRateBase>`. For performance reasons
+ * it is essential that derived classes use the keyword `final` to
+ * de-virtualize `virtual` methods.
+ *
+ * Methods defined for the abstract base class are not aware of specialized
+ * data handlers defined by the template class `ReactionRate<DataType>`
+ * and thus can be exposed to the API.
+ *
+ * @warning This class is an experimental part of the %Cantera API and
+ *    may be changed or removed without notice.
+ */
+class ReactionRateBase
+{
+public:
+    virtual ~ReactionRateBase() {}
+
+    //! Identifier of reaction type
+    virtual std::string type() const = 0;
+
+    //! Evaluate reaction rate based on temperature
+    virtual double eval(double T) const = 0;
+
+    //! Evaluate reaction rate based on temperature and pressure
+    virtual double eval(double T, double P) const = 0;
+
+    //! Evaluate reaction rate based on bulk phase
+    virtual double eval(const ThermoPhase& bulk) const = 0;
+
+    //! Evaluate reaction rate derivative based on temperature
+    virtual double ddT(double T) const = 0;
+
+    //! Evaluate reaction rate derivative based on temperature and pressure
+    virtual double ddT(double T, double P) const = 0;
+
+    //! Evaluate reaction rate derivative based on bulk phase
+    virtual double ddT(const ThermoPhase& bulk) const = 0;
+};
+
+
+//! Class template for reaction rate definitions with specialized DataType
+/**
+ * This class template ensures that derived objects are aware of specialized
+ * data types, which are passed by `MultiRateBase` evaluators.
+ *
+ * @warning This class is an experimental part of the %Cantera API and
+ *    may be changed or removed without notice.
+ */
+template <class DataType>
+class ReactionRate : public ReactionRateBase
+{
+public:
+    //! Constructor
+    ReactionRate() = default;
+
+    //! Update information specific to reaction
+    virtual void update(const DataType& shared_data, const ThermoPhase& bulk) {}
+
+    //! Evaluate reaction rate
+    virtual double eval(const DataType& shared_data) const = 0;
+
+    virtual double eval(double T) const override {
+        return eval(DataType(T));
+    }
+
+    virtual double eval(double T, double P) const override {
+        return eval(DataType(T, P));
+    }
+
+    virtual double eval(const ThermoPhase& bulk) const override {
+        return eval(DataType(bulk));
+    }
+
+    //! Evaluate derivative of reaction rate with respect to temperature
+    virtual double ddT(const DataType& shared_data) const {
```

Comment:
Clang gives me lots of warning about this kind of thing -- if you use `override` once in a class, you need to use it everywhere else in the class where it applies.
```suggestion
    virtual double ddT(const DataType& shared_data) const override {
```

Reply:
This one is actually not available to the base class, as `DataType` only becomes available in the template. I'll make sure that docstrings make this sufficiently clear.

Reply:
Ah, sorry, I guess I commented on the wrong one. I can confirm the warnings are now gone, though.